### PR TITLE
Link label security

### LINKED ISSUE #1493: Extremely slow processing on malformed markdown (0.6.2)
**Describe the bug**
~~Infinite loop~~ Extremely slow processing on malformed markdown

Edit: I tested it under Chrome and it finished significantly faster than Firefox, but still quite slow (7 seconds).

I believe it is a combination of malformed input and an exponential-runtime regexp combining to hang the parser.  I tried to strip the markdown to the exact syntax that breaks the parser, but it's difficult to narrow down.  The reason I suspect it's a run-away regexp is that removing obvious non-syntax like the strings of plain alphabetic words allows it to finish, but only after 30+ seconds of processing.

(for anyone curious, the markdown is from some old notes I had on the AREXX programming language)

**To Reproduce**

Parse this (be careful):
```
INDEX(string, pattern[, start)` : searches for the first occurrence of pattern in string, starting from start: `INDEX("123123", "23", 3)` == `5`
`INSERT(new, old[, start][, length][, pad])` : inserts the new string into the old string after the specified position (default is 0), new string is truncated or padded (default is " ") to the specified length, if start is beyond the end of old old will be padded
`LASTPOS(pattern, string[, start])` : searches backwards for the last occurrence of pattern in string, starting from start: `LASTPOS("123123", "23", 4)` == `2`
`LINES(file)` : returns the number of lines typed ahead at the interactive stream: `push("a line"); push("second line"); lines(STDIN); /* == 2 */`
`MAX(number, number[, number,...])` : obvious
`MIN(number, number[, number,...])` : obvious
`OPEN(filehandle, filename[, "APPEND"|"READ"|"WRITE"])` : opens file, returns boolean for success: `OPEN("MyCon", "CON:160/50/320/100/MyCon/CDS")` == `1`
`OVERLAY(new, old[, start][, length][, pad])` : overlays new string onto old one at start for length chars padding with pad if necessary: `OVERLAY("4", "123", 5, 5)` == `"123-4----"`
`POS(pattern, string[, start])` : same as index
```

[Test string at marked.js.org/demo](https://marked.js.org/demo/?text=INDEX(string%2C%20pattern%5B%2C%20start)%60%20%3A%20searches%20for%20the%20first%20occurrence%20of%20pattern%20in%20string%2C%20starting%20from%20start%3A%20%60INDEX(%22123123%22%2C%20%2223%22%2C%203)%60%20%3D%3D%20%605%60%0A%60INSERT(new%2C%20old%5B%2C%20start%5D%5B%2C%20length%5D%5B%2C%20pad%5D)%60%20%3A%20inserts%20the%20new%20string%20into%20the%20old%20string%20after%20the%20specified%20position%20(default%20is%200)%2C%20new%20string%20is%20truncated%20or%20padded%20(default%20is%20%22%20%22)%20to%20the%20specified%20length%2C%20if%20start%20is%20beyond%20the%20end%20of%20old%20old%20will%20be%20padded%0A%60LASTPOS(pattern%2C%20string%5B%2C%20start%5D)%60%20%3A%20searches%20backwards%20for%20the%20last%20occurrence%20of%20pattern%20in%20string%2C%20starting%20from%20start%3A%20%60LASTPOS(%22123123%22%2C%20%2223%22%2C%204)%60%20%3D%3D%20%602%60%0A%60LINES(file)%60%20%3A%20returns%20the%20number%20of%20lines%20typed%20ahead%20at%20the%20interactive%20stream%3A%20%60push(%22a%20line%22)%3B%20push(%22second%20line%22)%3B%20lines(STDIN)%3B%20%2F*%20%3D%3D%202%20*%2F%60%0A%60MAX(number%2C%20number%5B%2C%20number%2C...%5D)%60%20%3A%20obvious%0A%60MIN(number%2C%20number%5B%2C%20number%2C...%5D)%60%20%3A%20obvious%0A%60OPEN(filehandle%2C%20filename%5B%2C%20%22APPEND%22%7C%22READ%22%7C%22WRITE%22%5D)%60%20%3A%20opens%20file%2C%20returns%20boolean%20for%20success%3A%20%60OPEN(%22MyCon%22%2C%20%22CON%3A160%2F50%2F320%2F100%2FMyCon%2FCDS%22)%60%20%3D%3D%20%601%60%0A%60OVERLAY(new%2C%20old%5B%2C%20start%5D%5B%2C%20length%5D%5B%2C%20pad%5D)%60%20%3A%20overlays%20new%20string%20onto%20old%20one%20at%20start%20for%20length%20chars%20padding%20with%20pad%20if%20necessary%3A%20%60OVERLAY(%224%22%2C%20%22123%22%2C%205%2C%205)%60%20%3D%3D%20%60%22123-4----%22%60%0A%60POS(pattern%2C%20string%5B%2C%20start%5D)%60%20%3A%20same%20as%20index&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

I'm running 0.6.2 downloaded from jsdelivr (the header says "/npm/marked@0.6.2/lib/marked.js"), however it's apparently also broken in the demo.  I'm doing all this under Firefox 67 but it appears to hang Chrome, too.

**Expected behavior**
Totally screwy marked up text, but not exponential runtime.

### PR DESCRIPTION
**Pull request by UziTech**
**Marked version:** master

## Description

Fixes the ReDOS vulnerability in #1493

This does break a current test with a single backtick in a link label

```markdown
[the ` character](/url)
```

That markdown can be fixed by escaping the backtick

```markdown
[the \` character](/url)
```

Fixes #1493 

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

### CODE REVIEW

**Comment by UziTech** [comment]
If this PR is merged I will create a new issue about the single backtick so we can still track it.

**Comment by UziTech** [comment]
This is a temporary fix. Links will need to be parsed with a function to be fully spec compliant to allow the single backtick along with nested brackets like we are parsing the href.

File: test/specs/new/nested_square_link.md [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [the `]` character](/url)
 
-[the ` character](/url)
+[the \` character](/url)
```
Comment (styfle) [inline test/specs/new/nested_square_link.md]: So this PR is going to introduce a regression but fix a redos issue?

Comment (UziTech) [reply test/specs/new/nested_square_link.md]: yes

**Pull request merged**